### PR TITLE
Fixes found with -ffreestanding disabled

### LINKED ISSUE #45157: cmake: Use of -ffreestanding disables many useful optimizations and compiler warnings
Zephyr unconditionally enables the C compiler flag -ffreestanding which is only necessary when the environment does not include the standard C library. This may be necessary when using the minimal C library, but it's not helpful when using a POSIX compatible C library such as newlib or picolibc.

Not only does -freestanding disable many builtin functions, it also disables compile-time warnings about buffer overflows for libc functions, disables optimizations based on internal compiler knowledge of libc semantics and generally makes code less secure and less performant.

Disabling this option isn't something that can be done casually though; any code which redefines libc functions, or which uses libc functions in ways that the compiler may think could cause a buffer overflow will generate compile errors or warnings. Which makes this a project policy issue, not just a bug in the build system...

### PR DESCRIPTION
**Pull request by keith-packard**
Here's a collection of fixes to drivers, subsys, tests and samples to resolve compiler warnings and errors found when -ffreestanding is taken out of the compiler command line, leaving it able to detect problems when using standard C library functions.

I think there are a couple of real bugs found here, although most of the changes relate to buffers sized for numbers not expected to span the full range of their declared type.

This series is no longer stacked on top of the picolibc-module series so printk/cbprintf errors will not be checked until those patches are also merged.

Closes: #45157 

Depends on #44096 (includes commits from #44096)

### CODE REVIEW

**Comment by stephanosio** [comment]
Rebased

**Comment by keith-packard** [comment]
> Rebased

Thanks; lemme get it actually passing all of the tests before we rebase; it's sitting on top of two pending PRs and I don't want to rebase those until reviews have finished to avoid confusing reviewers. It's been interesting finding all of the places clang discovers issues with the current code...

File: tests/lib/sprintf/src/main.c [hunk 1]
```diff
@@ -764,8 +778,10 @@ void test_fprintf(void)
 	ret = fprintf(stdout, "");
 	zassert_equal(ret, 0, "fprintf failed!");
 
+#ifndef __PICOLIBC__	/* this is UB */
```
Comment (evgeniy-paltsev) [inline tests/lib/sprintf/src/main.c:781]: Hmm, we are checking against `CONFIG_PICOLIBC` in one place and against `__PICOLIBC__` in other places in same file. I'm wondering if it is intended?

Comment (keith-packard) [reply tests/lib/sprintf/src/main.c]: This patch is part of PR #44096, and I've fixed that to use CONFIG_PICOLIBC consistently.

**Review by andyross — DISMISSED** [review]
Scanned all the code I recognize well enough to review and all looks great.  This +1 will be wiped out with the next rev of course, but putting it here anyway so there's a historical record.

**Comment by keith-packard** [comment]
I rebased this back to main so it can be reviewed independently of the picolibc merge.

**Review by nvlsianpu — APPROVED** [review]
+1 to the flash driver

**Review by dcpleung — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Pull request merged**